### PR TITLE
feat(benchpress): initial support for firefox

### PR DESCRIPTION
**Pull request by hankduan**
[Review on Reviewable](https://reviewable.io/reviews/angular/angular/2419)

### CODE REVIEW

File: modules/benchpress/src/firefox_extension/lib/perf_profile_helper.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+};
+
+/**
+ * Reads the perf profile at the specified path and returns the processed
+ * events.
+ * @param {string} perfProfilePath The path of the perf profile.
+ * @return {Array<EventRecord>} An array of recognized events that are captured
+ *     within the perf profile.
+ */
+exports.readPerfProfile = function(perfProfilePath) {
```
Comment (tbosch) [inline modules/benchpress/src/firefox_extension/lib/perf_profile_helper.ts]: Can we read the perf profile without writing it to a file?

Comment (tbosch) [inline modules/benchpress/src/firefox_extension/lib/perf_profile_helper.ts]: This functionality should all go into a `FirefoxDriverExtension`, similar to `ChromeDriverExtension`. However, for this you should change the firefox extension to:

1. report the profile data without writing it to a file (i.e. directly as result of the JS call).
2. separate stopping the profiler from reading the profile, as Benchpress will read the profile multiple times (after every iteration and maybe later even in between). I.e. the profile would be started when benchpress creates the `FirefoxDriverExtension` and not stopped until benchpress is done (or maybe actually never stopped, this is the behavior in Chrome...)

**Comment by hankduan** [comment]
All comments addressed. PTAL @tbosch

File: modules/benchpress/src/firefox_extension/lib/perf_profile_helper.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+var EventRecord = function(eventName, startTime, endTime) {
```
Comment (tbosch) [inline modules/benchpress/src/firefox_extension/lib/perf_profile_helper.ts]: Hank,
the mapping of a FF event to an event that we use in benchpress should not be done in the Firefox extension. The idea is to keep the firefox extension as stable as possible.

Please move the code for `EventRecord.categorizeEvent` into a new class `FirefoxWebdriverExtension`, similar and next to `ChromeWebDriverExtension`.

File: modules/benchpress/test/firefox_extension/spec.ts [hunk 3]
```diff
@@ -3,16 +3,25 @@
 /// <reference path="../../../angular2/typings/jasmine/jasmine.d.ts" />
 
 var fs = require('fs');
+var perfProfileHelper = require('../../src/firefox_extension/lib/perf_profile_helper.js');
```
Comment (tbosch) [inline modules/benchpress/test/firefox_extension/spec.ts]: Additional to this test, please also create a full integration e2e test that uses benchpress directly with your new `FirefoxWebdriverExtension` (see above).

Comment (hankduan) [inline modules/benchpress/test/firefox_extension/spec.ts]: Can you point me to the e2e test that the chrome driver is using? I can only find https://github.com/angular/angular/blob/master/modules/benchpress/test/webdriver/chrome_driver_extension_spec.ts, but that's not exactly an e2e test as it's using a mock driver adapter.

**Comment by hankduan** [comment]
PTAL

File: modules/benchpress/src/firefox_extension/lib/main.ts [hunk 4]
```diff
@@ -1,52 +1,158 @@
...
+  addStartEvent(name: string, timeStarted: number) {
+    this._markerEvents.push({ph: 'b', ts: timeStarted - this._profilerStartTime, name: name});
+  }
+
+  addEndEvent(name: string, timeEnded: number) {
+    this._markerEvents.push({ph: 'e', ts: timeEnded - this._profilerStartTime, name: name});
+  }
+}
 
 function saveToFile(savePath: string, body: string) {
```
Comment (tbosch) [inline modules/benchpress/src/firefox_extension/lib/main.ts]: Do we still need this? I thought we are not storing the profile in files any more...

File: modules/benchpress/src/firefox_extension/lib/main.ts [hunk 5]
```diff
@@ -1,52 +1,158 @@
...
+    perfEvents.sort(function(event1, event2) { return event1.ts - event2.ts; });  // Sort by ts
+    return perfEvents;
+  }
+
+  /**
+   * @param {Object} perfProfile The perf profile JSON object.
+   * @return {Array<Object>} An array of recognized events that are captured
+   *     within the perf profile.
+   */
+  _convertPerfProfileToEvents(perfProfile: any): List<any> {
```
Comment (tbosch) [inline modules/benchpress/src/firefox_extension/lib/main.ts]: Sweet, you also added frame timing!

However, I think we really need a set of unit tests for this logic here.
You could move the logic into a separate file that does not `require` anything and test input/output.

File: modules/benchpress/src/firefox_extension/lib/main.ts (hunk 4 above)
Comment (hankduan) [inline modules/benchpress/src/firefox_extension/lib/main.ts]: We don't. I kept it there since it's super useful for debugging purposes.

**Comment by tbosch** [comment]
And the PR title / commit message seems off.
Maybe something with `feat(benchpress): initial support for firefox`

File: scripts/ci/test_e2e_js.sh [hunk 6]
```diff
@@ -28,7 +28,7 @@ if [[ -n "$E2E_BROWSERS" ]]; then
   OPTIONS="--browsers=$E2E_BROWSERS";
 fi
 
-./node_modules/.bin/protractor protractor-js.conf.js $OPTIONS
-./node_modules/.bin/protractor protractor-js.conf.js $OPTIONS --benchmark --dryrun
+# ./node_modules/.bin/protractor protractor-js.conf.js $OPTIONS
```
Comment (tbosch) [inline scripts/ci/test_e2e_js.sh]: Can you enable this again?

**Pull request closed without merging**